IRGen: Fix multi-payload enum lowering

aschwaighofer opened the pull request:
When we have a private resilient enum that is resilient because one of
its payloads is resilient but we have disabled resilience in the
context of lowering the enum as a class member (sigh), we must consider
it's payload's layout enum in the minimal domain (ignore the private
visibility) because we don't truly know the layout.

rdar://41308521

aschwaighofer committed:
IRGen: Fix multi-payload enum lowering

When we have a private resilient enum that is resilient because one of
its payloads is resilient but we have disabled resilience in the
context of lowering the enum as a class member (sigh), we must consider
it's payload's layout enum in the minimal domain (ignore the private
visibility) because we don't truly know the layout.

rdar://41308521

aschwaighofer commented:
@swift-ci Please test

slavapestov commented on lib/IRGen/GenEnum.cpp:
```diff
@@ -5258,6 +5258,12 @@ EnumImplStrategy::get(TypeConverter &TC, SILType type, EnumDecl *theEnum) {
+    // If we forced completely fragile layout we need to check the
+    // fixed-sized'ness in the minimal scope. Otherwise, we get wrong answer for
+    // private resilient enum types that have an resilient payload.
+    layoutScope = ResilienceExpansion::Minimal;
```
Perhaps this should be part of TC.IGM.getResilienceExpansionForLayout()? Are there other places where we call it and get it wrong?

aschwaighofer replied:
I sunk the logic.

aschwaighofer committed:
Sink the logic into getResilienceExpansionForLayout

This clearly satisfies the postcondition that 'Calling isResilient() with this scope will always return false.'

aschwaighofer committed:
Try to fix i368 failure

aschwaighofer commented:
@swift-ci Please test

aschwaighofer committed:
Fix compile error

aschwaighofer commented:
@swift-ci Please test

aschwaighofer committed:
Try to fix test

aschwaighofer commented:
@swift-ci Please test

aschwaighofer commented:
Oh, right this test case is not going to work on linux where we don't have to expand resilient types in classes because of objc interrupt ...

aschwaighofer committed:
Don't run test on linux - it does not apply

aschwaighofer committed:
Fix test again

aschwaighofer commented:
@swift-ci Please test
